### ISSUE TITLE
Whether to support multi-threaded calls？

### ISSUE THREAD

**Issue by wenminluo** [usnistgov/REFPROP-issues#122]
Hello,
   Can the exported function in "Refprop.dll" support multi-threaded calls?If not, is there anyone who has implemented it in a special way?

**Comment by ianhbell** [comment]
REFPROP is not thread-safe, so multi-threaded calls into REFPROP are not possible.  Furthermore, REFPROP reads from files in a non-thread-safe (and also not process-safe) way, so you need to fork separate threads to call REFPROP, and alsoo may need to make copies of all the FLUIDS and MIXTURES for each process.  An example of calling REFPROP in parallel is here: https://github.com/usnistgov/REFPROP-wrappers/blob/master/wrappers/python/ctypes/examples/multiprocessing_example.py

**Comment by wenminluo** [comment]
> 
> 
> REFPROP is not thread-safe, so multi-threaded calls into REFPROP are not possible. Furthermore, REFPROP reads from files in a non-thread-safe (and also not process-safe) way, so you need to fork separate threads to call REFPROP, and alsoo may need to make copies of all the FLUIDS and MIXTURES for each process. An example of calling REFPROP in parallel is here: https://github.com/usnistgov/REFPROP-wrappers/blob/master/wrappers/python/ctypes/examples/multiprocessing_example.py

Ok,Thanks!

**Comment by paarfi** [comment]
I had the same question some years back. If you look at the Refprop code, you can get an idea of how difficult this would be. It might almost need to be rewritten from scratch. It would be nice if RefProp were threadsafe, but given how many new refrigerant blends are coming on the market, the continuous improvement that the RefProp team is always doing, and their very limited resources, I'd expect a task this large is going to be a lower priority for them.

I ended creating interpolation tables based on calls to Refprop, and then writing my own thread-safe calls to use those tables. You can get a good start on what's needed to do that by looking at what Ian had done in his CoolProp project, which you can find here: https://github.com/CoolProp/CoolProp

**Comment by EricLemmon** [comment]
Refprop has the ability to make multiple calls in a threaded process when using the same fluid or mixture.  It is not too complicated to set up when calling the Fortran code directly.  We've never tried it by calling the DLL, but it is most likely possible.  However, Ian's method that he gave above is by far the preferred method for most, if not all, users.

**Comment by wenminluo** [comment]
> 
> 
> I had the same question some years back. If you look at the Refprop code, you can get an idea of how difficult this would be. It might almost need to be rewritten from scratch. It would be nice if RefProp were threadsafe, but given how many new refrigerant blends are coming on the market, the continuous improvement that the RefProp team is always doing, and their very limited resources, I'd expect a task this large is going to be a lower priority for them.
> 
> I ended creating interpolation tables based on calls to Refprop, and then writing my own thread-safe calls to use those tables. You can get a good start on what's needed to do that by looking at what Ian had done in his CoolProp project, which you can find here: https://github.com/CoolProp/CoolProp

This gives me valuable ideas.Thanks!

**Comment by wenminluo** [comment]
> 
> 
> Refprop has the ability to make multiple calls in a threaded process when using the same fluid or mixture. It is not too complicated to set up when calling the Fortran code directly. We've never tried it by calling the DLL, but it is most likely possible. However, Ian's method that he gave above is by far the preferred method for most, if not all, users.

Thank you for your reply！

**Comment by ianhbell** [comment]
It is not possible to make the calls to the DLL thread-safe in the way that REFPROP is configured right now.  _Don't do it!_  Rather, you need to fork separate processes and probably also copy all the fluid files for each process.
